File: MessengerApp/app/src/main/java/com/example/messengerapp/AES.java
package com.example.messengerapp;

import java.nio.charset.StandardCharsets;
import java.security.KeyPair;
import java.security.KeyPairGenerator;
import java.security.PrivateKey;
import java.security.PublicKey;
import java.security.SecureRandom;
import java.util.Scanner;

import javax.crypto.Cipher;
// import javax.xml.bind.DatatypeConverter;

public class AES {

    private static final String RSA = "RSA";
    private static Scanner sc;

    // Generating public & private keys
    // using RSA algorithm.
    public static KeyPair generateRSAKkeyPair()
            throws Exception
    {
        SecureRandom secureRandom = new SecureRandom ();
        KeyPairGenerator keyPairGenerator = KeyPairGenerator.getInstance(RSA);

        keyPairGenerator.initialize(2048, secureRandom);
        return keyPairGenerator.generateKeyPair();
    }

    // Encryption function which converts
    // the plainText into a cipherText
    // using private Key.
    public static byte[] do_RSAEncryption(String plainText, PublicKey publicKey) throws Exception
    {
        Cipher encrypt=Cipher.getInstance("RSA/ECB/PKCS1Padding");
        encrypt.init(Cipher.ENCRYPT_MODE, publicKey);
        byte[] encryptedMessage = encrypt.doFinal(plainText.getBytes(StandardCharsets.UTF_8));
        return encryptedMessage;
    }

    // Decryption function which converts
    // the ciphertext back to the
    // orginal plaintext.
    public static String do_RSADecryption(byte[] cipherText, PrivateKey privateKey) throws Exception
    {
        Cipher decrypt=Cipher.getInstance("RSA/ECB/PKCS1Padding");
        decrypt.init(Cipher.DECRYPT_MODE, privateKey);
        String decryptedMessage = new String(decrypt.doFinal(cipherText), StandardCharsets.UTF_8);
        return decryptedMessage;
    }
}

File: MessengerApp/app/src/main/java/com/example/messengerapp/AES.kt
package com.example.messengerapp

import android.os.Build
import android.support.annotation.RequiresApi
import java.nio.charset.StandardCharsets
import java.security.*
import java.security.interfaces.ECPublicKey
import java.security.spec.PKCS8EncodedKeySpec
import java.security.spec.X509EncodedKeySpec
import java.util.*
import javax.crypto.Cipher


object AES {
    private val RSA = "RSA"

    fun generateRSAKkeyPair():KeyPair {
        val secureRandom = SecureRandom()
        val keyPairGenerator = KeyPairGenerator.getInstance(RSA)
        keyPairGenerator.initialize(2048, secureRandom)
        return keyPairGenerator.generateKeyPair()
    }

    fun do_RSAEncryption(plainText: String, publicKey: PublicKey) : String {
        val encrypt = Cipher.getInstance("RSA/ECB/PKCS1Padding")
        encrypt.init(Cipher.ENCRYPT_MODE, publicKey)
        return Base64.getEncoder().encodeToString(encrypt.doFinal(plainText.toByteArray(StandardCharsets.UTF_8)))
    }


    fun do_RSADecryption(cipherText: String, privateKey: PrivateKey) : String {
        val decrypt = Cipher.getInstance("RSA/ECB/PKCS1Padding")
        decrypt.init(Cipher.DECRYPT_MODE, privateKey)
        return String(decrypt.doFinal(  Base64.getDecoder().decode( cipherText)  ), StandardCharsets.UTF_8)
    }

    fun publicKeyToString(publicKey: PublicKey) : String {
        return Base64.getEncoder().encodeToString(publicKey.encoded)
    }

    fun stringToPublicKey(encodedString: String) : PublicKey {
        val publicBytes: ByteArray = Base64.getDecoder().decode(encodedString)
        val keySpec = X509EncodedKeySpec(publicBytes)
        val keyFactory = KeyFactory.getInstance("RSA")
        return keyFactory.generatePublic(keySpec)
    }

    fun privateKeyToString(privateKey: PrivateKey) : String {
        return Base64.getEncoder().encodeToString(privateKey.encoded)
    }

    fun stringToPrivateKey(encodedString: String) : PrivateKey {
        val privateBytes: ByteArray = Base64.getDecoder().decode(encodedString)
        val keySpec = PKCS8EncodedKeySpec(privateBytes)
        val keyFactory = KeyFactory.getInstance("RSA")
        return keyFactory.generatePrivate(keySpec)
    }
}